pmlc/target/x86/stencil_tpp_binary.cc

```diff
@@ -0,0 +1,219 @@
+// Copyright 2020 Intel Corporation
+
+#include <string>
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+
+#include "pmlc/dialect/pxa/ir/matchers.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/autotile.h"
+#include "pmlc/dialect/pxa/transforms/stencil.h"
+#include "pmlc/dialect/pxa/transforms/tile.h"
+#include "pmlc/dialect/stdx/ir/ops.h"
+#include "pmlc/target/x86/pass_detail.h"
+#include "pmlc/target/x86/passes.h"
+#include "pmlc/util/logging.h"
+#include "pmlc/util/matchers.h"
+#include "pmlc/util/util.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::target::x86 {
+
+namespace pxa = dialect::pxa;
+namespace stdx = dialect::stdx;
+
+namespace {
+
+struct TppOperand {
+  Value memref;
+  AffineMap accessMap;
+  AffineMap tileMap;
+};
+
+template <typename TOp>
+Optional<TppOperand> getTppOperand(TOp op, Block *block,
+                                   ArrayRef<BlockArgument> idxs,
+                                   SmallVectorImpl<Value> &mapOperands) {
+  Optional<pxa::RelativeAccessPattern> rap =
+      pxa::computeRelativeAccess(op, block);
+  if (!rap)
+    return None;
+
+  AffineValueMap outerValueMap =
+      pxa::convertToValueMap(op.getContext(), rap->outer);
+  mapOperands.append(outerValueMap.getOperands().begin(),
+                     outerValueMap.getOperands().end());
+
+  AffineMap tileMap = pxa::makeTileMap(op.getContext(), op.getAffineMap(),
+                                       op.getMapOperands(), idxs);
+
+  return TppOperand{op.getMemRef(), outerValueMap.getAffineMap(), tileMap};
+}
+bool isLocallyDefined(AffineParallelOp &op, Value &source) {
+  if (!source.isa<BlockArgument>()) {
+    // If the definition of load's source is in "op", it is too complex to
+    // stencil
+    auto defOp = source.getDefiningOp();
+    while (!isa<FuncOp>(defOp)) {
+      if (defOp == op.getOperation())
+        return true;
+      defOp = defOp->getParentOp();
+    }
+  }
+  return false;
+}
+
+class StencilImpl : public pxa::StencilBase {
+private:
+  std::string opName;
+  template <typename OpTy>
+  void maybeCaptureGeneric(Optional<pxa::StencilCapture> &capture,
+                           const std::string &inName) {
```

Comment:
Use `llvm::StringRef` (which I think is re-exported in `mlir::StringRef`).

Reply:
Thanks, corrected